src/bsoncxx/document/value.cpp

```diff
@@ -23,6 +23,9 @@ namespace bsoncxx {
 BSONCXX_INLINE_NAMESPACE_BEGIN
 namespace document {
 
+value::value() noexcept : _data(NULL, [](std::uint8_t*) {}) {
```

Comment:
I went with optional, because it made the move and copy operators much nicer to implement.  If we decide to make document::value default constructible anyway, then let's do it in a separate commit.